core/kernel/shared/sprintf.rb

```diff
@@ -0,0 +1,945 @@
+        format("%.10e", 109.52).should == "1.0952000000e+02"
+        format("%.10E", 109.52).should == "1.0952000000E+02"
+        format("%.10f", 10.952).should == "10.9520000000"
+        format("%.10a", 196).should == "0x1.8800000000p+7"
+        format("%.10A", 196).should == "0X1.8800000000P+7"
+      end
+
+      # strange behavior
+      it "does not affect G format" do
+        format("%.10g", 12.1234).should == "12.1234" # expected "12.1234000000"
+        format("%.10g", 123456789).should == "123456789" # expected "1.2345700000e+08"
+      end
+    end
+
+    context "string formats" do
+      it "determines the maximum number of characters to be copied from the string" do
+        format("%.1p", [1]).should == "["
+        format("%.2p", [1]).should == "[1"
+        format("%.10p", [1]).should == "[1]"
+        format("%.0p", [1]).should == ""
+
+        format("%.1s", "abc").should == "a"
+        format("%.2s", "abc").should == "ab"
+        format("%.10s", "abc").should == "abc"
+        format("%.0s", "abc").should == ""
+      end
+    end
+  end
+
+  describe "reference by name" do
+    describe "%<name>s style" do
+      it "uses value passed in a hash argument" do
+        format("%<foo>d", foo: 123).should == "123"
+      end
+
+      it "supports flags, width, precision and type" do
+        format("%+20.10<foo>f", foo: 10.952).should == "      +10.9520000000"
+      end
+
+      it "allows to place name in any position" do
+        format("%+15.5<foo>f", foo: 10.952).should == "      +10.95200"
+        format("%+15<foo>.5f", foo: 10.952).should == "      +10.95200"
+        format("%+<foo>15.5f", foo: 10.952).should == "      +10.95200"
+        format("%<foo>+15.5f", foo: 10.952).should == "      +10.95200"
+      end
+
+      it "cannot be mixed with unnamed style" do
+        -> () {
+          format("%d %<foo>d", 1, foo: "123")
+        }.should raise_error(ArgumentError)
+      end
+
+      it "raises KeyError when there is no matching key" do
+        -> () {
+          format("%<foo>s", {})
+        }.should raise_error(KeyError)
+      end
+    end
+
+    describe "%{name} style" do
+      it "uses value passed in a hash argument" do
+        format("%{foo}", foo: 123).should == "123"
+      end
+
+      it "does not support type style" do
+        format("%{foo}d", foo: 123).should == "123d"
+      end
+
+      it "supports flags, width and precision" do
+        format("%-20.5{foo}", foo: "123456789").should == "12345               "
+      end
+
+      it "cannot be mixed with unnamed style" do
+        -> () {
+          format("%d %{foo}", 1, foo: "123")
+        }.should raise_error(ArgumentError)
+      end
+
+      it "raises KeyError when there is no matching key" do
+        -> () {
+          format("%{foo}", {})
+        }.should raise_error(KeyError)
+      end
+
+      # strange behavior
+      # expected to call to_str and only then to_s
```

Comment:
I think that's OK, see my comment above about to_s/to_str.

Reply:
:+1: